sklearn/ensemble/_hist_gradient_boosting/splitting.pyx

```diff
@@ -638,18 +727,43 @@ cdef inline Y_DTYPE_C _split_gain(
     """
     cdef:
         Y_DTYPE_C gain
-    gain = negative_loss(sum_gradient_left, sum_hessian_left,
-                         l2_regularization)
-    gain += negative_loss(sum_gradient_right, sum_hessian_right,
-                          l2_regularization)
-    gain -= negative_loss_current_node
```

Comment:
For reviewers:

The gain computation has been updated to:

- first compute values of left and right child
- cap those values according to the bounds for monotonic constraints
- discard any split that does not respect left < right (INC) or right < left (DEC)
- compute the loss reduction from the previously (bounded) computed values